lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+        } else if (keystr_len >= 5 && memcmp(keystr, "HTTP_", 5) == 0) {
+            value = h2o_mruby_to_str(mrb, value);
+            h2o_mruby_split_header_pair(ctx->shared, key, value, handle_header_env_key, &subreq->super);
+        } else if (keystr_len != 0){
+            /* set to req->env */
+            value = h2o_mruby_to_str(mrb, value);
+            h2o_vector_reserve(&super->pool, &super->env, super->env.size + 2);
+            super->env.entries[super->env.size] = h2o_strdup(&super->pool, keystr, keystr_len);
+            super->env.entries[super->env.size + 1] = h2o_strdup(&super->pool, RSTRING_PTR(value), RSTRING_LEN(value));
+            super->env.size += 2;
+        }
+    }
+#undef RETRIEVE_ENV
+#undef CALC_HASH
+#undef CHECK_KEY
+
+    /* do validations */
+#define CHECK_REQUIRED(k, v, non_empty) do { \
+    if (mrb_nil_p(v)) { \
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "missing required environment key: " k)); \
+        goto Failed; \
+    } else if (non_empty && RSTRING_LEN(v) == 0) { \
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, k " must be not empty")); \
+        goto Failed; \
+    } \
+} while (0)
+    CHECK_REQUIRED("REQUEST_METHOD", method, 1);
+    CHECK_REQUIRED("rack.url_scheme", scheme, 1);
+    CHECK_REQUIRED("SCRIPT_NAME", script_name, 0);
+    CHECK_REQUIRED("PATH_INFO", path_info, 0);
+    CHECK_REQUIRED("QUERY_STRING", query_string, 0);
+#undef CHECK_REQUIRED
+
+    if (RSTRING_LEN(script_name) != 0 && RSTRING_PTR(script_name)[0] != '/') {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "SCRIPT_NAME must start with `/`"));
+        goto Failed;
+    }
+    if (RSTRING_LEN(path_info) != 0 && RSTRING_PTR(path_info)[0] != '/') {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "PATH_INFO must start with `/`"));
+        goto Failed;
+    }
+    if (mrb_nil_p(http_host) && (mrb_nil_p(server_name) || mrb_nil_p(server_port))) {
+        mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "HTTP_HOST or (SERVER_NAME and SERVER_PORT) is required"));
+        goto Failed;
+    }
+
+    if (!is_reprocess) {
+        /* ensure that SCRIPT_NAME is not modified */
+        h2o_iovec_t confpath = ctx->handler->pathconf->path;
+        size_t confpath_len_wo_slash = confpath.base[confpath.len - 1] == '/' ? confpath.len - 1 : confpath.len;
+        if (!(RSTRING_LEN(script_name) == confpath_len_wo_slash && memcmp(RSTRING_PTR(script_name), confpath.base, confpath_len_wo_slash) == 0)) {
+            mrb->exc = mrb_obj_ptr(mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "can't modify `SCRIPT_NAME` with `H2O.next`. Is `H2O.reprocess` what you want?"));
+            goto Failed;
+        }
+    }
+
+#define STR_TO_IOVEC(val) h2o_iovec_init(RSTRING_PTR(val), RSTRING_LEN(val))
+
+    /* construct url and parse */
+    h2o_iovec_t *url_comps = alloca(sizeof(*url_comps) * 9);
```

Comment:
Why not just use `url_comps[9]`? Use of `alloca` is only necessary when the number of elements is unknown at compile-time.